File: src/main/java/com/icloud/observer/after/ChatServer.java
package com.icloud.observer.after;

import java.util.*;

public class ChatServer {

    private Map<String, List<Subscriber>> subscribers = new HashMap<>();

    public void register(String subject, Subscriber subscriber) {
        if (subscribers.containsKey(subject)) {
            subscribers.get(subject).add(subscriber);
        } else {
            var list = new ArrayList<Subscriber>();
            list.add(subscriber);
            this.subscribers.put(subject, list);
        }
    }

    public void unregister(String subject, Subscriber subscriber) {
        if (subscribers.containsKey(subject)) {
            subscribers.get(subject).remove(subscriber);
        }
    }

    public void sendMessage(User user, String subject, String message) {
        if (this.subscribers.containsKey(subject)) {
            String userMessage = user.getName() + " : " + message;
            this.subscribers.get(subject).forEach(s -> s.handleMessage(userMessage));
        }
    }

}

File: src/main/java/com/icloud/observer/review/ChatServer.kt
package com.icloud.observer.review

typealias Subject = String


class ChatServer {

    private var subscribers: HashMap<Subject, ArrayList<Subscriber>> = HashMap()

    fun register(subject: Subject, subscriber: Subscriber) {
        if (this.subscribers.hasSubject(subject)) {
            this.subscribers.addSubscriber(subject, subscriber)
        } else {
            this.subscribers.addNewSubjectSubscriber(subject, subscriber)
        }
    }

    fun unregister(subject: Subject, subscriber: Subscriber) {
        if (this.subscribers.hasSubject(subject)) {
            this.subscribers[subject]?.remove(subscriber)
        }
    }

    fun sendMessage(user: User, subject: Subject, message: String) {
        if (this.subscribers.hasSubject(subject)) {
            this.subscribers[subject]?.forEach { s: Subscriber ->
                val userMessage = "from ${user.name} : $message\nto ${(s as User).name}"
                s.handleMessage(userMessage)
            }
        }
    }


    private fun HashMap<Subject, ArrayList<Subscriber>>.hasSubject(subject: Subject): Boolean {
        return this.containsKey(subject)
    }

    private fun HashMap<Subject, ArrayList<Subscriber>>.addSubscriber(subject: Subject, subscriber: Subscriber) {
        this[subject]?.add(subscriber)
    }

    private fun HashMap<Subject, ArrayList<Subscriber>>.addNewSubjectSubscriber(
        subject: Subject,
        subscriber: Subscriber
    ) {
        this[subject] = (this[subject] ?: ArrayList())
        this[subject]!!.add(subscriber)
    }
}